### ISSUE TITLE
Xwayland-satellite broken on master

### ISSUE THREAD

**Issue by sodiboo** [NixOS/nixpkgs#352725]
### Describe the bug
On latest nixpkgs master commit, `xwayland-satellite` is broken.

At time of writing, the latest commit to `nixos-unstable` is https://github.com/NixOS/nixpkgs/commit/807e9154dcb16384b1b765ebe9cd2bba2ac287fd. If the steps below don't work in a few days from now, replace `nixos-unstable` in commands with https://github.com/NixOS/nixpkgs/commit/807e9154dcb16384b1b765ebe9cd2bba2ac287fd. The latest commit to master as i'm writing is https://github.com/NixOS/nixpkgs/commit/769a6a566120266b06f86e6c0436ee2b60c0238b.

- Working commit (current `nixos-unstable`): https://github.com/NixOS/nixpkgs/commit/807e9154dcb16384b1b765ebe9cd2bba2ac287fd
- Broken commit (current `master`): https://github.com/NixOS/nixpkgs/commit/769a6a566120266b06f86e6c0436ee2b60c0238b

### Steps To Reproduce
1. `nix run github:nixos/nixpkgs/nixos-unstable#xwayland-satellite -- :12`
2. Observe that it says `Connected to Xwayland on :12`
3. `nix run github:nixos/nixpkgs/master#xwayland-satellite -- :12`
4. Observe that Xwayland never initializes

### Expected behavior
Xwayland should initialize and xwayland-satellite should connect to it within less than a second of startup. (from what i can observe, usually within less than 100ms)

### Screenshots
Here is some terminal output showing several xwayland-satellite packages. The first two invocations are identical, as i have the `xwayland-satellite-unstable` package from my niri flake installed in my system. Those packages are used here to indicate that this isn't necessarily a problem with nixpkgs' package expression, it is inherently a problem with a dependency (likely Xwayland). They're built against `nixos-unstable`.

![image](https://github.com/user-attachments/assets/287d5ede-2546-4479-b09d-776bb9793303)

### Additional context
- This issue is almost certainly caused by https://github.com/NixOS/nixpkgs/pull/352289, which is [not yet in nixos-unstable](https://nixpk.gs/pr-tracker.html?pr=352289)

This doesn't seem to be caused by us. I think it needs to be fixed upstream. I'm making this issue first, though, in case it is actually nixos-specific.

### Notify maintainers

@if-loop69420 @getchoo are the other maintainers of the `xwayland-satellite` package

### Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.54, NixOS, 24.11 (Vicuna), 24.11.20241029.807e915`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.9`
 - channels(root): `"nixos"`
 - channels(sodiboo): `""`
 - nixpkgs: `/nix/store/m68ikm8045fj7ys7qvgr608z9l70hh1k-source`

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by sodiboo** [comment]
The PR updating `xwayland` was backported to `nixos-24.05`, and although that branch has no `xwayland-satellite` to break, my niri flake has a package that works on `nixos-24.05` and when i try to check if the same regression applies there, i see no issue. This suggests the issue may not be caused by an `xwayland` update after all, since the backported PR (https://github.com/NixOS/nixpkgs/pull/352314) does not break an `xwayland-satellite` built against `nixos-24.05`

![image](https://github.com/user-attachments/assets/9f2ff7ab-765c-4882-9525-7b5237f6fa3b)

**Comment by sodiboo** [comment]
I'm currently bisecting nixpkgs to find the offending commit in case it's not that PR updating xwayland (which it seems not to be, given that it doesn't break on 24.05), This is taking a while, as Hydra doesn't cache every commit to nixpkgs, and i'm currently on the third iteration of a bisect script which has taken over am hour so far and doesn't seem to be anywhere near the end. Will update once this bisect yields a result, or if i fall asleep before that point, hopefully i'll have an update in the morning if it finished by then

**Comment by sodiboo** [comment]
Yeah uhh i'm still bisecting because i've tried several times and it just leads me to Entirely The Wrong Commit. so:..? if anyone else is seeing this, feel free to try to bisect it to find the actual cause, you might get there way sooner than me.

**Comment by sodiboo** [comment]
Yeah, so this is most definitely not an issue with Xwayland, but rather with upgrading mesa.

After writing my previous comment, i noticed that the reason my bisects were failing were always because `rust-cbindgen` was broken. So i wrote a script that checks `xwayland-satellite` after applying [this patch](https://github.com/NixOS/nixpkgs/commit/89798d91509343438392208b97ce1f913d9a6fd6.patch), and that led me to https://github.com/NixOS/nixpkgs/commit/09de22541f3c2c8389362de4fea2cca72953c3f0, which i discovered nearly 30 hours ago now.

<details>

<summary>
Here's the bisect script i used.
</summary>

```fish
#!/usr/bin/env fish

git apply cbindgen.patch
nom build .#xwayland-satellite
git reset --hard

if test $status -ne 0
    exit 125 # skip
end

./result/bin/xwayland-satellite 2> log &

sleep 5

kill $last_pid

pkill -9 Xwayland


bat log | rg "Connected to Xwayland"
```

</details>

- https://github.com/NixOS/nixpkgs/pull/349114

The reason the same backported changes didn't break `nixos-24.05` is because that branch doesn't update mesa.

I also noticed that my system is on `24.2.4`, which is equal to the old version (working xwayland-satellite) and lower than the new version (broken xwayland-satellite). See attached image:

![image](https://github.com/user-attachments/assets/5fd261ed-9e9d-4f8e-8c52-68a1dc5c70e8)

As such, i figured "oh, is it just because the package is newer than my system? what if i build with `mesa 24.2.5`?".

Turns out, a lot of things depend on mesa, and [my computer didn't like that rebuild very much](https://gaysex.cloud/notes/a02s0mmrn7mk00n8). That rebuild exceeded 100 GB memory usage the first time i tried, and when i limited the amount of jobs to hopefully reduce memory usage, it did run for longer before locking up. In fact, it still hasn't locked up, but [electron has been building for more than 24 hours](https://gaysex.cloud/notes/a04gduugn7mk012q). I think maybe this build isn't going to finish anytime soon, if ever, so i actually have absolutely no idea if upgrading mesa on my system would break it. Since i'm impatient, i'm just gonna stop waiting for it, and report my findings so far, without testing for a system with a matching mesa version.

That leaves me with the disappointing result: i'm not sure if this is just a "mesa version mismatch" thing, or the more interesting "this mesa version is actually bad".

**Comment by getchoo** [comment]
See my [other comment](https://github.com/NixOS/nixpkgs/pull/354326#issuecomment-2465105609)

I can't check if this is a Mesa issue since I have only have a NVIDIA card; I'm hesitant to say it is though, as there are no upstream bug reports of this to my knowledge

>  Turns out, a lot of things depend on mesa, and [my computer didn't like that rebuild very much](https://gaysex.cloud/notes/a02s0mmrn7mk00n8).

You may want to try [`system.replaceRuntimeDependencies`](https://nixos.org/manual/nixos/stable/options.html#opt-system.replaceRuntimeDependencies)

```nix
{ pkgs, inputs, ... }:
let
  nixpkgs-master = inputs.nixpkgs-master.legacyPackages.${pkgs.stdenv.hostPlatform.system};
in
{
  system.replaceRuntimeDependencies = [
    { original = pkgs.mesa; replacement = nixpkgs-master.mesa; }
    { original = pkgs.i686Linux.mesa; replacement = nixpkgs-master.i686Linux.mesa; }
  ];
}
```

**Comment by sodiboo** [comment]
I realize now also that either way the issue isn't caused by upgrading this package anyways. Like, the old version is broken already. I didn't open a PR because `broken -> broken` felt unproductive, but if you observe the package as working in its current state, then there's no reason to prolong this PR (since at any rate, it definitely doesn't make the package more broken *and* it does work for you)

**Comment by getchoo** [comment]
Is this still an issue?

**Comment by sodiboo** [comment]
I cannot easily reproduce the issue with the current versions of software installed on my system. I guess it was just a mesa mismatch thing. Closing, as it doesn't seem to affect anyone else (and doesn't affect me anymore, either), and given what i know about the root cause, it's very unlikely that we can even do anything about it anyhow.
